### DISCUSSION TITLE
Use Framework with Deno?

### DISCUSSION THREAD

**Discussion by charnould** [observablehq/framework#878] [Q&A] [answered]
Hi,
First, thanks for making the Framework: it looks fantastic to me.
For various reasons I use Deno runtime, and I was wondering if it's possible to run the Framework?
I'm currently not able to do so (nor install, nor run dev...).
I'm probably missing something but I can't figure it out: any clue/help?
Thanks.

**Comment by mythmon** [comment]
We haven't done any testing with Deno internally, so at this point I think you're the expert on Framework+Deno 🙂.

What kind of problems do you run into?

**Comment by mbostock** [comment]
Are you asking to write data loaders using Deno? Or for Framework itself to use Deno (for the build and preview commands, say)? Or perhaps you want to use Framework as an API (library) within Deno?

The first of these should be the easiest of the three. It should be readily doable to switch from Node to Deno for JavaScript (.js) and TypeScript (.ts) data loaders. The only challenge is that they both compete for the same file extensions. So either we’ll need the config file to declare a preference for Deno, or we’ll need some other hint such as a shebang that we should use a different interpreter. If this is what you’re interested in, do you have preference?

**Comment by nshiab** [comment, marked as answer]
I just tested the [Deno v2 release candidate](https://deno.com/blog/v2.0-release-candidate). It works without problems so far.

Instead of `npm run dev`, use `deno task dev`.

Same thing for the other commands.

**Comment by huw** [comment]
Hijacking this to add some tips for Deno data loaders:

1. In `observablehq.config.js`, add:

```js
export default {
  interpreters: {
    ".ts": ["deno", "--env-file", "--allow-env", "--allow-net", "--allow-read"]
  },
};
```

(Or whichever Deno flags you need for your project)

2. Add [`nodeModulesDir: manual`](https://docs.deno.com/runtime/fundamentals/configuration/#node-modules-directory) to your `package.json` to share packages across runtimes if you're using another package manager (which you should be doing, to run Framework):

```json
{
  "nodeModulesDir": "manual"
}
```

3. Don't use `Deno.stdout.write`, because it isn't guaranteed to output the whole buffer. This gave me hours of headaches for no reason. Instead, write:

```ts
const writer = Deno.stdout.writable.getWriter();
await writer.write(buffer);
await writer.close();
```

I'll try and add more as I run into them, but it's pretty seamless so far.
